classpath/java/io/BufferedInputStream.java

```diff
@@ -54,28 +59,7 @@ public int read(byte[] b, int offset, int length) throws IOException {
 
       count += remaining;
       position += remaining;
-      offset += remaining;
-      length -= remaining;
     }
-
-    while (length > 0) {
```

Comment:
Let me see if I can summarize:

You're pointing out (rightly) that the code is not well optimized for when `length < this.limit` (perhaps much less, as in your example of a single-byte read).  You're code also seems to handle that better.  On the other hand, the current code _does_ deal well with the case where `length > this.limit`, but unless I'm missing something, your code doesn't handle that well, since it forces a fixed upper limit on the amount of data that can be read per call to `read`.  I think both cases are important.

Does that make sense?  Is that what you're getting at?  If so, perhaps we can come up with a solution that solves both cases well.